### DISCUSSION TITLE
Grammar

### DISCUSSION THREAD

**Discussion by DarkNacho** [bucanero/apollo-ps3#24] [General]
Hi I want to write an interpreter for BSD scrip so I would like to know if you have the grammar for it, you have a lot of examples and also an implementation but I am not that good at C for understanding your implementation and extract myself the grammar from the code.
Of course the basic function, set variables and that is easy to understand it from the examples but I am really more interested in the checksums grammar. I would appreciate if you have one and you could share it, if you don't have one, an explanation in more details of how the checksum grammar works would be nice also. Thanks in advace

**Comment by bucanero** [comment]
well, I think there's no official definition or document about the BSD format. I guess it evolved as needed when they were doing cheats and patches for the PS3.

You can find some notes about the BSD format here (not complete, some commands missing):
https://github.com/bucanero/apollo-ps3/blob/master/docs/_EXAMPLE.ps3savepatch

The original Visual Basic parsing code from BSD is available in #7 . My code is basically my personal re-interpretation of the VB source. I know it's not 100% accurate (some commands are not implemented), but for most cases and tests it works fine. 

Checksums in general work by defining a calculation range (like from `0x00A0` to `0x0FFF`), the hash algorithm (like md5, crc32, sha1) and a variable to store the result. Then it usually saves that variable value back to the file at some specific offset (e.g. `0x0008`) because that's where the game will read the checksum value.
